src/base/static/state/ducks/map.js

```diff
@@ -1,311 +1,581 @@
+    .reduce((flat, toFlatten) => flat.concat(toFlatten), [])
+    .map(layer => layer.id);
+
+  return {
+    type: LOAD_STYLE_AND_METADATA,
+    payload: {
+      style,
+      sourcesMetadata,
+      layerGroupsMetadata,
+      basemapLayerIds,
+      interactiveLayerIds,
+    },
+  };
+}
+
+export const mapViewportPropType = PropTypes.shape({
+  height: PropTypes.number.isRequired,
+  width: PropTypes.number.isRequired,
+  latitude: PropTypes.number,
+  longitude: PropTypes.number,
+  zoom: PropTypes.number,
+  pitch: PropTypes.number,
+  bearing: PropTypes.number,
+  transitionInterpolator: PropTypes.object,
+  transitionDuration: PropTypes.number,
+  transitionEasing: PropTypes.func,
+  minZoom: PropTypes.number.isRequired,
+  maxZoom: PropTypes.number.isRequired,
+});
+
+export const mapStylePropType = PropTypes.shape({
+  version: PropTypes.number,
+  name: PropTypes.string,
+  sources: PropTypes.object,
+  layers: PropTypes.arrayOf(PropTypes.object),
+});
+
+// Reducers:
+// TODO(luke): refactor our current implementation in AppView to use
 const INITIAL_STATE = {
-  visibleBasemapId: undefined,
-  layerStatuses: {},
-  featureFilters: [],
-  updatingFilterGroupId: undefined,
-  updatingFilterTargetLayer: undefined,
-  mapPosition: undefined,
-  isMapSizeValid: true,
-  isMapDragging: false,
+  viewport: {
+    width: 0,
+    height: 0,
+    minZoom: 1,
+    maxZoom: 18,
+    transitionInterpolator: new FlyToInterpolator(),
+    transitionEasing: easeInOutCubic,
+  },
+  style: {
+    version: 8,
+    name: "Mapseed",
+    glyphs: "mapbox://fonts/mapbox/{fontstack}/{range}.pbf",
+    sprite: `${window.location.protocol}//${
+      window.location.host
+    }/static/css/images/markers/spritesheet`,
+    sources: {},
+    layers: [],
+  },
+  layerGroupsMetadata: {},
+  sourcesMetadata: {},
+  basemapLayerIds: [],
+  interactiveLayerIds: [],
+  isMapSizeValid: false,
   isMapDragged: false,
+  isMapDragging: false,
 };
 
 export default function reducer(state = INITIAL_STATE, action) {
-  let newStatuses;
-
   switch (action.type) {
-    case UPDATE_MAP_DRAGGING:
+    case UPDATE_GEOJSON_FEATURES:
       return {
         ...state,
-        isMapDragging: action.payload,
-      };
-    case SET_MAP_SIZE_VALIDITY:
-      return {
-        ...state,
-        isMapSizeValid: action.payload,
+        style: {
+          ...state.style,
+          sources: {
+            ...state.style.sources,
+            [action.payload.sourceId]: {
+              ...state.style.sources[action.payload.sourceId],
+              data: {
+                ...state.style.sources[action.payload.sourceId].data,
+                features:
+                  action.payload.mode === "replace"
+                    ? action.payload.newFeatures
+                    : state.style.sources[
+                        action.payload.sourceId
+                      ].data.features.concat(action.payload.newFeatures),
```

Comment:
In this case, aren't we *adding* geojson features? In that case, I think this should be calling a separate `CREATE` action, not an `UPDATE` action (ie `CREATE_GEOJSON_FEATURES`)

It might be a bit more code, but it would provide more clarity versus passing in `mode`.

Reply:
Yes, I think you're right about the semantics here 👍